crates/ra_hir_ty/src/_match.rs

```diff
@@ -0,0 +1,1174 @@
+                            };
+                        }
+
+                        if found_unimplemented {
+                            Err(MatchCheckNotImplemented)
+                        } else {
+                            Ok(Usefulness::NotUseful)
+                        }
+                    } else {
+                        let matrix = matrix.specialize_constructor(&cx, &constructor)?;
+                        let v = v.expand_wildcard(&cx, &constructor)?;
+
+                        is_useful(&cx, &matrix, &v)
+                    }
+                } else {
+                    Ok(Usefulness::NotUseful)
+                }
+            }
+            _ => {
+                // Either not all constructors are covered, or the only other arms
+                // are wildcards. Either way, this pattern is useful if it is useful
+                // when compared to those arms with wildcards.
+                let matrix = matrix.specialize_wildcard(&cx);
+                let v = v.to_tail();
+
+                is_useful(&cx, &matrix, &v)
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+/// Similar to TypeCtor, but includes additional information about the specific
+/// value being instantiated. For example, TypeCtor::Bool doesn't contain the
+/// boolean value.
+enum Constructor {
+    Bool(bool),
+    Tuple { arity: usize },
+    Enum(EnumVariantId),
+}
+
+/// Returns the constructor for the given pattern. Should only return None
+/// in the case of a Wild pattern.
+fn pat_constructor(cx: &MatchCheckCtx, pat: PatIdOrWild) -> MatchCheckResult<Option<Constructor>> {
+    let res = match pat.as_pat(cx) {
+        Pat::Wild => None,
+        Pat::Tuple(pats) => Some(Constructor::Tuple { arity: pats.len() }),
+        Pat::Lit(lit_expr) => match cx.body.exprs[lit_expr] {
+            Expr::Literal(Literal::Bool(val)) => Some(Constructor::Bool(val)),
+            _ => return Err(MatchCheckNotImplemented),
+        },
+        Pat::TupleStruct { .. } | Pat::Path(_) => {
+            let pat_id = pat.as_id().expect("we already know this pattern is not a wild");
+            let variant_id =
+                cx.infer.variant_resolution_for_pat(pat_id).ok_or(MatchCheckNotImplemented)?;
+            match variant_id {
+                VariantId::EnumVariantId(enum_variant_id) => {
+                    Some(Constructor::Enum(enum_variant_id))
+                }
+                _ => return Err(MatchCheckNotImplemented),
+            }
+        }
+        _ => return Err(MatchCheckNotImplemented),
+    };
+
+    Ok(res)
+}
+
+fn all_constructors_covered(
+    cx: &MatchCheckCtx,
+    constructor: &Constructor,
+    used_constructors: &[Constructor],
+) -> bool {
+    match constructor {
+        Constructor::Tuple { arity } => {
+            used_constructors.iter().any(|constructor| match constructor {
+                Constructor::Tuple { arity: used_arity } => arity == used_arity,
```

Comment:
since you assume the pattern type-checks, this check is superfluous, I think

Reply:
I agree. I've replaced this check with a `debug_assert_eq`.